include/circt/Conversion/StandardToHandshake.h

```diff
@@ -86,16 +123,35 @@ class HandshakeLowering {
       operands.push_back(startCtrl);
       rewriter.replaceOpWithNewOp<handshake::ReturnOp>(retOp, operands);
     }
+
+    // Store the number of block arguments in each block
+    DenseMap<Block *, unsigned> numArgsPerBlock;
+    for (auto &block : r.getBlocks())
+      numArgsPerBlock[&block] = block.getNumArguments();
+
+    // Apply SSA maximization on the newly added entry block argument to
+    // propagate it explicitly between the start-point of the control-only
+    // network and the function's terminators
+    if (failed(maximizeSSA(startCtrl, rewriter)))
+      return failure();
+
+    // Identify all block arguments belonging to the control-only network
+    // (needed to later insert control merges after those values). These are the
+    // last arguments to blocks that got a new argument during SSA maximization
+    for (auto &[block, numArgs] : numArgsPerBlock)
```

Comment:
Not sure if I understand this in its entirety, but aren't the control signals always at the end of the argument list? If so, why not just updating all of them as we are either way looping over all blocks?

Reply:
> but aren't the control signals always at the end of the argument list?

They are indeed, whenever they are added by SSA maximization. However, I do not know whether it is possible to have a function containing a block that wouldn't need a control argument. In that case, `maximizeSSA(startCtrl, rewriter)` wouldn't add an extra argument to the block and we couldn't assume that all blocks have an entry control. Hence this conservative approach which checks for the addition of a new argument explicitly.

Perhaps such blocks could only exist in "degenerate" code that is unreachable anyway (e.g., they couldn't dominate any block that has a `return`, otherwise they would need to forward the control signal) and that we don't care about supporting. What do you think?

Reply:
I'm not sure if this is an invariant that is guaranteed by the previous transformation step or not. If we aren't sure, it's better to be conservative, I guess.